File: ashgf/gradient/estimators.py
"""Gradient estimators: Gaussian smoothing, Gauss-Hermite quadrature, and
Lipschitz-constant estimation.

All critical loops are vectorised via NumPy to minimise Python-level
overhead in high-dimensional settings.
"""

from __future__ import annotations

import os
_os = os.environ.setdefault("OPENBLAS_NUM_THREADS", "1")
os.environ.setdefault("OMP_NUM_THREADS", "1")
os.environ.setdefault("MKL_NUM_THREADS", "1")

import logging
import math
import os
from concurrent.futures import ThreadPoolExecutor
from typing import Callable

import numpy as np

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Cache for Gauss-Hermite quadrature nodes & weights (computed once per m)
# ---------------------------------------------------------------------------
_GH_CACHE: dict[int, tuple[np.ndarray, np.ndarray, np.ndarray]] = {}


def _get_gauss_hermite(m: int) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Return (p_nodes, p_w, weights) for m-point Gauss-Hermite quadrature.

    Results are cached so that the eigenvalue decomposition of the
    tridiagonal Jacobi matrix is performed only once per distinct ``m``.
    """
    if m not in _GH_CACHE:
        p_nodes, weights = np.polynomial.hermite.hermgauss(m)
        p_w = p_nodes * weights  # pre-multiplied: w_k * p_k
        _GH_CACHE[m] = (p_nodes, p_w, weights)
    return _GH_CACHE[m]


# ---------------------------------------------------------------------------
# Parallel / batched evaluation of the objective function
# ---------------------------------------------------------------------------

# Number of threads to use when ``n_jobs`` is not explicitly given.
# Set via the environment variable ``ASHGF_N_JOBS``, defaulting to 1.
_DEFAULT_N_JOBS: int = int(os.environ.get("ASHGF_N_JOBS", "1"))


def _get_n_jobs(n_jobs: int | None = None) -> int:
    """Return the effective number of parallel workers."""
    if n_jobs is not None:
        return max(1, n_jobs)
    return max(1, _DEFAULT_N_JOBS)


def _parallel_eval(
    f: Callable[[np.ndarray], float],
    points: list[np.ndarray],
    *,
    n_jobs: int | None = None,
) -> list[float]:
    """Evaluate ``f`` on each point in ``points``, optionally in parallel.

    Parameters
    ----------
    f : callable
        The objective function.
    points : list of np.ndarray
        List of input points at which to evaluate ``f``.
    n_jobs : int or None
        Number of threads.  If ``None``, the value of the environment
        variable ``ASHGF_N_JOBS`` is used (default 1).

    Returns
    -------
    list of float
        ``f(p)`` for each ``p`` in ``points``, in the same order.
    """
    n = len(points)
    nj = _get_n_jobs(n_jobs)

    if nj <= 1 or n < 4:
        return [f(p) for p in points]

    # Parallel path — use executor.map which preserves order and avoids
    # the overhead of as_completed + dict bookkeeping.
    with ThreadPoolExecutor(max_workers=nj) as executor:
        return list(executor.map(f, points))


__all__ = [
    "gaussian_smoothing",
    "gauss_hermite_derivative",
    "estimate_lipschitz_constants",
]


# ---------------------------------------------------------------------------
# Gaussian smoothing
# ---------------------------------------------------------------------------


def gaussian_smoothing(
    x: np.ndarray,
    f: Callable[[np.ndarray], float],
    sigma: float,
    directions: np.ndarray,
    *,
    n_jobs: int | None = None,
) -> np.ndarray:
    """
    Central Gaussian Smoothing gradient estimator.

    g(x) ≈ (1/(2σ·M)) Σ_{i=1}^{M} [f(x+σ·d_i) - f(x-σ·d_i)] · d_i

    Parameters
    ----------
    x : np.ndarray, shape (d,)
        Point at which to estimate the gradient.
    f : callable
        Objective function f: R^d → R.
    sigma : float
        Smoothing bandwidth.
    directions : np.ndarray, shape (M, d)
        Matrix of random directions (one per row).
    n_jobs : int or None
        Number of threads for parallel evaluation.

    Returns
    -------
    grad : np.ndarray, shape (d,)
        Estimated gradient.
    """
    M = len(directions)

    # Pre-scale directions
    sigma_dirs = sigma * directions  # (M, d)
    dim = len(x)

    # Build all perturbed points via broadcasting (no Python loop)
    # all_arr[0::2] = x + sigma_dirs, all_arr[1::2] = x - sigma_dirs
    all_arr = np.empty((2 * M, dim))
    all_arr[0::2] = x[None, :] + sigma_dirs
    all_arr[1::2] = x[None, :] - sigma_dirs
    points = list(all_arr)  # list of (dim,) views

    # Evaluate (parallel if n_jobs > 1)
    results = _parallel_eval(f, points, n_jobs=n_jobs)

    # ---- Vectorised gradient assembly ----
    # diff[j] = f(x + σ·d_j) - f(x - σ·d_j)
    results_arr = np.asarray(results)  # (2M,)
    diff = results_arr[0::2] - results_arr[1::2]  # (M,)
    grad = np.dot(diff, directions) / (2.0 * sigma * M)  # (d,) = (M,) @ (M, d)

    return grad


# ---------------------------------------------------------------------------
# Gauss-Hermite quadrature
# ---------------------------------------------------------------------------


def gauss_hermite_derivative(
    x: np.ndarray,
    f: Callable[[np.ndarray], float],
    sigma: float,
    basis: np.ndarray,
    m: int,
    value_at_x: float | None = None,
) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """
    Estimate directional derivatives using Gauss-Hermite quadrature.

    For each basis direction b_i:
        D_i f(x) ≈ (2/(σ√π)) Σ_{k=1}^{m} w_k · p_k · f(x + σ·p_k·b_i)

    where (p_k, w_k) are the nodes and weights of the m-point
    Gauss-Hermite quadrature rule.

    Parameters
    ----------
    x : np.ndarray, shape (d,)
        Point at which to estimate.
    f : callable
        Objective function.
    sigma : float
        Smoothing bandwidth.
    basis : np.ndarray, shape (d, d)
        Orthonormal basis matrix (each row is a direction vector b_i).
    m : int
        Number of quadrature points (must be odd for central node at 0).
    value_at_x : float or None
        f(x). If None, it will be evaluated.

    Returns
    -------
    grad : np.ndarray, shape (d,)
        Estimated gradient (sum of directional derivatives times basis vectors).
    evals_matrix : np.ndarray, shape (d, m)
        Matrix of function evaluations: row i = evaluations along direction i.
    points : np.ndarray, shape (m,)
        Quadrature nodes (shared across all directions).
    derivatives : np.ndarray, shape (d,)
        Array of estimated directional derivatives D_i f(x).
    """
    dim = len(x)
    if value_at_x is None:
        value_at_x = f(x)

    p_nodes, p_w, _weights = _get_gauss_hermite(m)
    # p_w[mid] == 0 because p_nodes[mid] == 0
    sigma_p = sigma * p_nodes  # (m,)
    mid = m // 2

    # Pre-compute scale factor for the quadrature
    quad_scale = 2.0 / (sigma * np.sqrt(math.pi))

    # ---- Build all non-central perturbed points via broadcasting ----
    # Shape: (dim, m, dim) → select k != mid → (dim, m-1, dim) → flatten
    perturbed = x[None, None, :] + sigma_p[None, :, None] * basis[:, None, :]
    k_mask = np.ones(m, dtype=bool)
    k_mask[mid] = False
    flat_arr = perturbed[:, k_mask, :].reshape(-1, dim)
    flat_points = list(flat_arr)

    # ---- Evaluate all points (parallel if ASHGF_N_JOBS > 1) ----
    flat_results = _parallel_eval(f, flat_points)

    # ---- Build evaluation matrix: (d, m) ----
    # Fill central column with f(x); other entries from flat_results.
    evals_matrix = np.empty((dim, m))
    evals_matrix[:, mid] = value_at_x
    # Reshape flat results to (dim, m-1) and assign to non-mid columns
    flat_reshaped = np.asarray(flat_results).reshape(dim, m - 1)
    evals_matrix[:, k_mask] = flat_reshaped

    # ---- Compute directional derivatives (vectorised) ----
    # derivatives[i] = quad_scale · Σ_k p_w[k] · evals_matrix[i, k]
    # Since p_w[mid] == 0, value_at_x contributes nothing.
    derivatives = quad_scale * (evals_matrix @ p_w)  # (d,) = (d, m) @ (m,)

    # ---- Gradient reconstruction: ∇f = Σ_i D_i f · b_i = Bᵀ · D ----
    grad = basis.T @ derivatives  # (d,) = (d, d) @ (d,)

    # The ``points`` return is kept as the 1-D array of nodes for
    # compatibility with ``estimate_lipschitz_constants``.
    return grad, evals_matrix, p_nodes, derivatives


# ---------------------------------------------------------------------------
# Lipschitz-constant estimation
# ---------------------------------------------------------------------------


def estimate_lipschitz_constants(
    evaluations: np.ndarray,
    points: np.ndarray,
    sigma: float,
) -> np.ndarray:
    """
    Estimate directional Lipschitz constants from quadrature data.

    Implements the thesis formula (Eq. ``Lipschitz constants``)::

        L_j = max_{{i,k} in I} |F(x+σ·p_i·ξ_j) - F(x+σ·p_k·ξ_j)| / (σ·|p_i - p_k|)

    where the index set ``I`` excludes pairs that are symmetric about
    the central quadrature node::

        I = {{ {i,k} | |i - floor(m/2)| != |k - floor(m/2)| }}

    Parameters
    ----------
    evaluations : np.ndarray, shape (d, m)
        Matrix of function evaluations along each direction (row = direction,
        column = quadrature node).
    points : np.ndarray, shape (m,)
        Quadrature nodes (shared across all directions).
    sigma : float
        Smoothing bandwidth.

    Returns
    -------
    lipschitz : np.ndarray, shape (d,)
        Estimated Lipschitz constants per direction.
    """
    d, m = evaluations.shape
    mid = m // 2

    # ---- Build all unordered pairs (i, k) with i < k ----
    i_idx, k_idx = np.triu_indices(m, k=1)  # each shape (P,)

    # ---- Exclude symmetric pairs around the centre ----
    dist_i = np.abs(i_idx - mid)
    dist_k = np.abs(k_idx - mid)
    keep = dist_i != dist_k

    i_idx = i_idx[keep]
    k_idx = k_idx[keep]
    # P ~ m*(m-1)/2 - floor((m-1)/2)  valid pairs

    if len(i_idx) == 0:
        return np.zeros(d)

    # ---- Vectorised ratios across all directions and all valid pairs ----
    # diff_evals: shape (d, P)
    diff_evals = np.abs(evaluations[:, i_idx] - evaluations[:, k_idx])

    # diff_pts: shape (P,)
    diff_pts = sigma * np.abs(points[i_idx] - points[k_idx])

    # ratios: (d, P) / (P,) -> (d, P)
    ratios = diff_evals / diff_pts[None, :]

    # Max over valid pairs per direction -> (d,)
    lipschitz = np.max(ratios, axis=1)

    return lipschitz

File: src/gradient/estimators.rs
//! Gradient estimators: Gaussian smoothing, Gauss-Hermite quadrature,
//! and Lipschitz-constant estimation.
//!
//! Uses the `gauss-quad` crate for quadrature nodes and weights
//! (probabilist's convention, weight `e^{-x²/2}`), avoiding the
//! `√2` factor ambiguity present in the Python reference.

use ndarray::{Array1, Array2};
use std::collections::HashMap;
use std::f64::consts::{PI, SQRT_2};
use std::sync::Mutex;

use crate::utils::parallel::parallel_eval;
use once_cell::sync::Lazy;

// ---------------------------------------------------------------------------
// Cache for Gauss-Hermite quadrature
// ---------------------------------------------------------------------------

/// Cached (nodes, p_w, weights) for a given number of points `m`.
static GH_CACHE: Lazy<Mutex<HashMap<usize, (Array1<f64>, Array1<f64>, Array1<f64>)>>> =
    Lazy::new(|| Mutex::new(HashMap::new()));

/// Get `(p_nodes, p_w, weights)` for `m`-point Gauss-Hermite quadrature.
///
/// `gauss-quad` provides physicist's nodes/weights (weight `e^{-x²}`).
/// We convert to the probabilist's convention (weight `e^{-x²/2}`):
///   v_k = √2 · p_k,   w_k^{prob} = √2 · w_k^{phys}
///
/// Returns:
/// - `p_nodes`: probabilist's quadrature nodes `v_k`
/// - `p_w`: pre-multiplied `v_k * w_k` (for the derivative formula)
/// - `weights`: probabilist's weights `w_k`
fn get_gauss_hermite(m: usize) -> (Array1<f64>, Array1<f64>, Array1<f64>) {
    let mut cache = GH_CACHE.lock().unwrap();
    if let Some(entry) = cache.get(&m) {
        return entry.clone();
    }

    let quad = gauss_quad::hermite::GaussHermite::init(m);
    // Convert physicist's → probabilist's
    let p_nodes = Array1::from_vec(quad.nodes) * SQRT_2;
    let w = Array1::from_vec(quad.weights) * SQRT_2;
    let p_w = &p_nodes * &w;

    let result = (p_nodes.clone(), p_w, w);
    cache.insert(m, result.clone());
    result
}

// ---------------------------------------------------------------------------
// Gaussian smoothing (central finite difference)
// ---------------------------------------------------------------------------

/// Estimate the gradient via central Gaussian smoothing.
///
/// `g ≈ (1/(2σ·M)) Σ_j [f(x+σ·d_j) - f(x-σ·d_j)] · d_j`
///
/// Parameters
/// ----------
/// x : Array1<f64>
///     Point at which to estimate the gradient.
/// f : &dyn Fn(&Array1<f64>) -> f64 + Sync
///     Objective function.
/// sigma : f64
///     Smoothing bandwidth.
/// directions : Array2<f64>
///     Matrix of directions (rows).
/// n_jobs : usize
///     Number of parallel threads.
///
/// Returns
/// -------
/// grad : Array1<f64>
///     Estimated gradient vector.
pub fn gaussian_smoothing<F>(
    x: &Array1<f64>,
    f: &F,
    sigma: f64,
    directions: &Array2<f64>,
    n_jobs: usize,
) -> Array1<f64>
where
    F: Fn(&Array1<f64>) -> f64 + Sync + ?Sized,
{
    let m = directions.nrows();
    let _dim = x.len();

    let sigma_dirs = directions.mapv(|v| sigma * v);

    // Build all perturbed points (interleaved + and -)
    let mut points: Vec<Array1<f64>> = Vec::with_capacity(2 * m);
    for j in 0..m {
        let d = sigma_dirs.row(j).to_owned();
        points.push(x + &d);
        points.push(x - &d);
    }

    let results = parallel_eval(f, &points, n_jobs);
    let results_arr = Array1::from_vec(results);

    // diff[j] = f(x + σ·d_j) - f(x - σ·d_j)
    let diff =
        results_arr.slice(ndarray::s![0..;2]).to_owned() - &results_arr.slice(ndarray::s![1..;2]);

    // grad = Σ diff_j · d_j / (2σM)
    let grad = diff.dot(directions) / (2.0 * sigma * m as f64);

    grad
}

// ---------------------------------------------------------------------------
// Gauss-Hermite quadrature
// ---------------------------------------------------------------------------

/// Estimate directional derivatives using Gauss-Hermite quadrature.
///
/// For each basis direction `b_i` (row of `basis`):
///
/// `D_i f(x) ≈ (1/σ)·(1/√(2π)) Σ_k w_k · v_k · f(x + σ·v_k·b_i)`
///
/// where `(v_k, w_k)` are the nodes and weights of the m-point
/// Gauss-Hermite quadrature (probabilist's convention).
///
/// Parameters
/// ----------
/// x : Array1<f64>
///     Point at which to estimate.
/// f : &dyn Fn(&Array1<f64>) -> f64 + Sync
///     Objective function.
/// sigma : f64
///     Smoothing bandwidth.
/// basis : Array2<f64>
///     Orthonormal basis matrix (each row is a direction).
/// m : usize
///     Number of quadrature points (must be odd).
/// value_at_x : Option<f64>
///     `f(x)`. If `None`, it will be evaluated.
///
/// Returns
/// -------
/// grad : Array1<f64>
///     Estimated gradient `∇f(x)`.
/// evals_matrix : Array2<f64>
///     Function evaluations (d × m): row i = evaluations along direction i.
/// nodes : Array1<f64>
///     Quadrature nodes (shared across all directions).
/// derivatives : Array1<f64>
///     Estimated directional derivatives `D_i f(x)`.
pub fn gauss_hermite_derivative<F>(
    x: &Array1<f64>,
    f: &F,
    sigma: f64,
    basis: &Array2<f64>,
    m: usize,
    value_at_x: Option<f64>,
    n_jobs: usize,
) -> (Array1<f64>, Array2<f64>, Array1<f64>, Array1<f64>)
where
    F: Fn(&Array1<f64>) -> f64 + Sync + ?Sized,
{
    assert!(m % 2 == 1, "m must be odd for central node at zero");
    let dim = x.len();
    let fx = value_at_x.unwrap_or_else(|| f(x));

    let (p_nodes, p_w, _weights) = get_gauss_hermite(m);
    let mid = m / 2;

    // Scale factor: (1/σ)·(1/√(2π))
    let quad_scale = 1.0 / (sigma * (2.0 * PI).sqrt());

    // ---- Build all non-central perturbed points ----
    let sigma_nodes = sigma * &p_nodes;
    let n_perturbed = dim * (m - 1);

    let mut points: Vec<Array1<f64>> = Vec::with_capacity(n_perturbed);
    for i in 0..dim {
        let bi = basis.row(i);
        for k in 0..m {
            if k == mid {
                continue;
            }
            points.push(x + &(sigma_nodes[k] * &bi));
        }
    }

    let flat_results = parallel_eval(f, &points, n_jobs);

    // ---- Build evaluation matrix: (d, m) ----
    let mut evals_matrix = Array2::<f64>::zeros((dim, m));
    evals_matrix.column_mut(mid).fill(fx);
    for i in 0..dim {
        for (j, k) in (0..m).filter(|&k| k != mid).enumerate() {
            evals_matrix[(i, k)] = flat_results[i * (m - 1) + j];
        }
    }

    // ---- Compute directional derivatives ----
    // D_i = quad_scale · Σ_k p_w[k] · evals[i, k]
    // Since p_w[mid] == 0 (v_mid == 0), fx contributes nothing.
    let derivatives = quad_scale * evals_matrix.dot(&p_w);

    // ---- Gradient reconstruction: ∇f = B^T · D ----
    let grad = basis.t().dot(&derivatives);

    (grad, evals_matrix, p_nodes, derivatives)
}

// ---------------------------------------------------------------------------
// Lipschitz-constant estimation
// ---------------------------------------------------------------------------

/// Estimate directional Lipschitz constants from quadrature data.
///
/// Implements the thesis formula:
///
/// `L_j = max_{i,k ∈ I} |F(x+σ·p_i·ξ_j) - F(x+σ·p_k·ξ_j)| / (σ·|p_i - p_k|)`
///
/// where `I` excludes pairs symmetric about the central node:
/// `I = {{i,k} : |i - ⌊m/2⌋| ≠ |k - ⌊m/2⌋|}`.
///
/// Parameters
/// ----------
/// evaluations : Array2<f64>, shape (d, m)
///     Function evaluations along each direction.
/// nodes : Array1<f64>, shape (m,)
///     Quadrature nodes.
/// sigma : f64
///     Smoothing bandwidth.
///
/// Returns
/// -------
/// lipschitz : Array1<f64>, shape (d,)
///     Estimated Lipschitz constants per direction.
pub fn estimate_lipschitz_constants(
    evaluations: &Array2<f64>,
    nodes: &Array1<f64>,
    sigma: f64,
) -> Array1<f64> {
    let (_d, m) = evaluations.dim();
    let mid = m / 2;

    // Build all unordered pairs (i, k) with i < k, excluding symmetric pairs
    let mut pairs: Vec<(usize, usize)> = Vec::new();
    for i in 0..m {
        for k in (i + 1)..m {
            let di = if i >= mid { i - mid } else { mid - i };
            let dk = if k >= mid { k - mid } else { mid - k };
            if di != dk {
                pairs.push((i, k));
            }
        }
    }

    if pairs.is_empty() {
        return Array1::zeros(_d);
    }

    // Vectorised over directions and pairs
    let mut lipschitz = Array1::<f64>::zeros(_d);
    for d_idx in 0.._d {
        let evals_row = evaluations.row(d_idx);
        let mut max_ratio: f64 = 0.0;
        for &(i, k) in &pairs {
            let diff_evals = (evals_row[i] - evals_row[k]).abs();
            let diff_pts = sigma * (nodes[i] - nodes[k]).abs();
            if diff_pts > 1e-14 {
                let ratio = diff_evals / diff_pts;
                max_ratio = max_ratio.max(ratio);
            }
        }
        lipschitz[d_idx] = max_ratio;
    }
    lipschitz
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::gradient::sampling::random_orthogonal;
    use crate::utils::SeededRng;
    use approx::assert_abs_diff_eq;

    /// Sphere: f(x) = ‖x‖², ∇f = 2x
    fn sphere(x: &Array1<f64>) -> f64 {
        x.dot(x)
    }

    #[test]
    fn gaussian_smoothing_sphere() {
        let x = Array1::from_vec(vec![1.0, 2.0, 3.0]);
        let n_runs = 1000;
        let mut avg_grad = Array1::zeros(3);
        for seed in 0..n_runs {
            let mut rng = SeededRng::new(seed);
            let dirs = crate::gradient::sampling::compute_directions(3, &mut rng);
            let grad = gaussian_smoothing(&x, &sphere, 1e-6, &dirs, 1);
            avg_grad = &avg_grad + &grad;
        }
        avg_grad /= n_runs as f64;
        let expected = 2.0 * &x;
        for i in 0..3 {
            assert_abs_diff_eq!(avg_grad[i], expected[i], epsilon = 3e-1);
        }
    }

    #[test]
    fn gh_derivative_sphere() {
        let x = Array1::from_vec(vec![1.0, 2.0, 3.0]);
        let mut rng = SeededRng::new(42);
        let basis = random_orthogonal(3, &mut rng);
        let (grad, _, _, _derivs) = gauss_hermite_derivative(&x, &sphere, 0.1, &basis, 5, None, 1);
        let expected = 2.0 * &x;
        // GH should be more accurate than finite diff
        for i in 0..3 {
            assert_abs_diff_eq!(grad[i], expected[i], epsilon = 1e-4);
        }
    }

    #[test]
    fn lipschitz_sphere_is_positive() {
        let x = Array1::zeros(3);
        let mut rng = SeededRng::new(42);
        let basis = random_orthogonal(3, &mut rng);
        let (_, evals, nodes, _) = gauss_hermite_derivative(&x, &sphere, 0.1, &basis, 5, None, 1);
        let lips = estimate_lipschitz_constants(&evals, &nodes, 0.1);
        // For sphere, Lipschitz constant ≈ 2
        for l in lips.iter() {
            assert!(l.is_finite() && *l > 0.0);
        }
    }
}
